securedrop/tests/functional/conftest.py

```diff
@@ -0,0 +1,158 @@
+from dataclasses import dataclass
+from multiprocessing.context import Process
+from pathlib import Path
+from typing import Generator
+import requests
+from selenium.webdriver.firefox.webdriver import WebDriver
+
+import pytest
+from models import Journalist
+from tests.functional.db_session import get_database_session
+from tests.functional.factories import SecureDropConfigFactory
+from tests.functional.sd_config_v2 import SecureDropConfig
+from tests.functional.web_drivers import WebDriverTypeEnum, get_web_driver
+import socket
+import time
+
+
+# Function-scoped so that tests can be run in parallel if needed
+@pytest.fixture(scope="function")
+def firefox_web_driver() -> WebDriver:
+    with get_web_driver(web_driver_type=WebDriverTypeEnum.FIREFOX) as web_driver:
+        yield web_driver
+
+
+# Function-scoped so that tests can be run in parallel if needed
+@pytest.fixture(scope="function")
+def tor_browser_web_driver() -> WebDriver:
+    with get_web_driver(web_driver_type=WebDriverTypeEnum.TOR_BROWSER) as web_driver:
+        yield web_driver
+
+
+def _get_unused_port() -> int:
+    s = socket.socket()
+    s.bind(("127.0.0.1", 0))
+    port = s.getsockname()[1]
+    s.close()
+    return port
+
+
+def _start_source_server(port: int, config_to_use: SecureDropConfig) -> None:
+    # This function will be called in a separate Process that runs the source app
+    # Modify the sdconfig module in the app's memory so that it mirrors the supplied config
+    # Do this BEFORE importing any other module of the application so the modified config is
+    # what eventually gets imported by the app's code
+    import sdconfig
+
+    sdconfig.config = config_to_use  # type: ignore
+
+    # Then start the source app
+    from source_app import create_app
+
+    source_app = create_app(config_to_use)  # type: ignore
+    source_app.run(port=port, debug=True, use_reloader=False, threaded=True)
+
+
+def _start_journalist_server(port: int, config_to_use: SecureDropConfig) -> None:
+    # This function will be called in a separate Process that runs the journalist app
+    # Modify the sdconfig module in the app's memory so that it mirrors the supplied config
+    # Do this BEFORE importing any other module of the application so the modified config is
+    # what eventually gets imported by the app's code
+    import sdconfig
+
+    sdconfig.config = config_to_use  # type: ignore
+
+    # Then start the journalist app
+    from journalist_app import create_app
+
+    journalist_app = create_app(config_to_use)  # type: ignore
+    journalist_app.run(port=port, debug=True, use_reloader=False, threaded=True)
+
+
+@dataclass(frozen=True)
+class SdServersFixtureResult:
+    source_app_base_url: str
+    journalist_app_base_url: str
+
+
+@contextmanager
+def spawn_sd_servers(
+    config_to_use: SecureDropConfig
+) -> Generator[SdServersFixtureResult, None, None]:
+    """Spawn the source and journalist apps as separate processes with the supplied config."""
```

Comment:
The above is the "killer" feature of the new `spawn_sd_servers` fixture and other factory code: the source and journalist apps to be spawned will use the supplied config_to_use regardless of the content of `securedrop/config.py`.

The confusion that comes from having to deal with live SDConfig objects that don't match the content of securedrop/config.py has been one of the biggest challenges for me when updating the test suite. This also comes from having different parts of the code sometimes use the live object or sometimes securedrop/config.py (it also depends on the order of imports in the code base...).

I think one medium term goal (started with some of the changes here) is to get rid of securedrop/config.py in the context of the test suite.